### PR TITLE
[KED-2603] - Create plotly preview

### PR DESCRIPTION
**Pull request by rashidakanchwala**
## Description

This PR is the first part of the Plotly Visualisation ticket. In this PR, we show Plotly preview when the user clicks on a Plotly Dataset node. 

## Development notes

To run this, you will need to copy paste the latest mock.animals.json to main and also add the node_plot.json in the nodes folder and rename the file to 'c3p345ed' (which is the node-id)

## QA notes

npm t ran successfully 

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [x] Added tests to cover my changes

## Legal notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

File: src/components/metadata/styles/metadata.scss
Comment (bru5) [inline src/components/metadata/styles/metadata.scss:257]: I think these colours should be replaced with their variables too. This one could probably just be solid black unless there's a need?

Comment (bru5) [inline src/components/metadata/styles/metadata.scss:257]: I think these colours should be replaced with their variables too. This one could probably just be solid black unless there's a need?

Comment (bru5) [inline src/components/metadata/styles/metadata.scss:257]: I think all the colours used in this file should be replaced with their global variables from `_variables.scss`. 

Also this one could probably just be solid black unless there's a need?

File: src/components/metadata/styles/metadata.scss
Comment (bru5) [inline src/components/metadata/styles/metadata.scss:271]: Does this need to account for both themes?

File: src/components/plotly-chart/index.js
Comment (bru5) [inline src/components/plotly-chart/index.js:8]: Think missing a newline here after the imports.

File: src/components/plotly-chart/index.js
Comment (bru5) [inline src/components/plotly-chart/index.js:16]: I think this variable might be better named more general as `config` or `plotConfig` (as I can see more things potentially added later for example).

File: src/components/plotly-chart/index.js
Comment (bru5) [inline src/components/plotly-chart/index.js:28]: I think with hooks `defaultProps` is no longer?

Instead I think write this like:

`const PlotlyChart = ({ theme, data={}, layout={} }) => {`

But that said - it looks like the default handling logic might just be removed from here and moved in to `<Plot>` instead as it seems more likely its responsibility to handle that?

File: src/components/plotly-chart/index.js
Comment (bru5) [inline src/components/plotly-chart/index.js:34]: Check the style guide but `light_preview` and `dark_preview` should be `LightPreview` and `DarkPreview`.

Also consider naming with a little more detail on what this is e.g. `LightPreviewTemplate`

File: src/config.js
Comment (bru5) [inline src/config.js:45]: Removing the `modularpipeline` flag should be another PR I think as it seems unrelated? If so revert this here I think.

File: src/reducers/nodes.js
Comment (bru5) [inline src/reducers/nodes.js:63]: Is this feature now removed or replaced then?

File: src/selectors/edges.js
Comment (bru5) [inline src/selectors/edges.js:105]: `getEdgesWithParams`

File: src/selectors/edges.js
Comment (bru5) [inline src/selectors/edges.js:108]: `nodesList`

File: src/selectors/edges.js
Comment (bru5) [inline src/selectors/edges.js:113]: I think this block needs formatting, it's strange that Prettier didn't format this?

File: src/setupTests.js
Comment (bru5) [inline src/setupTests.js:10]: Sorry to nitpick but please use a space before the start of the comment so it's a little easier to skim read

File: src/store/__mocks__/load-data.js
Comment (bru5) [inline src/store/__mocks__/load-data.js:5]: I'm not sure why these are not already using camel case, but would you mind updating all of them for consistency if possible?

File: src/utils/plot-templates/dark.js
Comment (bru5) [inline src/utils/plot-templates/dark.js:1]: Could you add a description of this file and where the format comes from in a comment block at the top? 
Even if it's just a link to Plotly docs would help I think.

**Review by bru5 — CHANGES_REQUESTED** [review]

File: src/components/metadata/index.js
Comment (studioswong) [inline src/components/metadata/index.js:38]: this could be simplied as `
metadata?.datasetType  ||   metadata?.node.type`

File: src/components/metadata/index.js
Comment (studioswong) [inline src/components/metadata/index.js:65]: looking at the icon assignment here it does look confusing - it would be best to rename the `icon` variable assignment name to something more declarative, like `nodeTypeIcon`

File: src/components/metadata/metadata.test.js
Comment (studioswong) [inline src/components/metadata/metadata.test.js:291]: Other than the above generic tests to test the data, please add in new tests to test the existence of the plot in the metadata panel, such as testing if certain classes related to the plot exists, etc.

File: src/config.js
Comment (studioswong) [reply src/config.js:45]: agree with @bru5 , not sure why the modular pipeline flag is removed here? Is this something done unintentionally from a merge conflict? 

Also speaking of, I think a feature flag needs to be set up for this feature?

File: src/reducers/nodes.js
Comment (studioswong) [reply src/reducers/nodes.js:63]: indeed it does seem strange to remove `datasetType` nodes as it still exists with `plot` type as an additional subset of dataset nodes?

File: src/selectors/edges.js
Comment (studioswong) [inline src/selectors/edges.js:105]: It is a bit weird for this selector to be added in the context of this PR as it doesn't seem to relate to the plotly feature - I also noticed that this selector is not used in any of the newly added code in this PR as well. I wonder what this selector is used for?

File: src/utils/data/animals.mock.json
Comment (studioswong) [inline src/utils/data/animals.mock.json:160]: Given that this dataset is also used with the dataset for the kedro-viz server, adding new fields to this file might cause the python tests to fail and will likely need updates to their test files. You might need to liase with the core team / whoever is maintaining the server codebase (cc @limdauto ) to check on this.

**Review by studioswong — CHANGES_REQUESTED** [review]
Thank you @rashidakanchwala ! 

One important thing: the plotly feature will require a series of changes to the kedro-viz server to accompany / enable this feature - I know that the relevant changes for this will be implemented in the new server, and given this case, this should be either be merged after [this ticket ](https://github.com/quantumblacklabs/kedro-viz/pull/432) is merged, or merged against the feature branch for this ticket, so we can test this feature alongside the new kedro-viz server on an actual response from the kedro-viz server to ensure it works in reality.   Otherwise it is still developed based on assumptions for the nodes endpoint which might cause problems down the line.

Also this feature should be hidden behind a feature flag ;)

File: src/selectors/edges.js
Comment (rashidakanchwala) [reply src/selectors/edges.js:113]: there's some problem. This was an old function in the previous parameters ticket which I had ommitted. I don't know how it got passed on. I will remove this entirely again

File: src/selectors/edges.js
Comment (rashidakanchwala) [reply src/selectors/edges.js:105]: yes, this is an old selector which I removed ages ago from my last ticket related to parameters. I have no clue, how it got added. It's not in the main. :S

File: src/utils/data/animals.mock.json
Comment (limdauto) [reply src/utils/data/animals.mock.json:160]: Please feel free to modify these files to test the frontend. We don't rely on them for the backend anymore.

**Comment by limdauto** [comment]
@studioswong @rashidakanchwala correct me if I'm wrong but if we merge this in and the backend doesn't return any Plotly-related data, the frontend will work exactly the same as before right? If that's the case, when this is ready, why can't it be merged to main? 

We can test it with the new backend in another branch.

**Review by rashidakanchwala — COMMENTED** [review]
> @studioswong @rashidakanchwala correct me if I'm wrong but if we merge this in and the backend doesn't return any Plotly-related data, the frontend will work exactly the same as before right? If that's the case, when this is ready, why can't it be merged to main?
> 
> We can test it with the new backend in another branch.

Yes, it won't be a problem. This will only work if a node with datasetType = plotly.PlotlyDataset is passed. I think it should be fine without a flag. @studioswong

**Comment by limdauto** [comment]
@rashidakanchwala I have talked to InfoSec. They are okay with us ignoring the new vulnerability discovered by Snyk here. So feel free to merge the PR when it's ready. After merging you can ask @yetudada to ignore the warning on `main`. There is no way for me to ignore it on a branch :(

File: .circleci/config.yml
Comment (limdauto) [inline .circleci/config.yml:289]: @rashidakanchwala this fixes the `windows_38` build problem.

File: package.json
Comment (limdauto) [inline package.json]: @rashidakanchwala this was causing some issue. I think you introducded it so I reverted it back.

File: src/selectors/edges.js
Comment (limdauto) [inline src/selectors/edges.js]: @rashidakanchwala These are unused variables so I removed them to pass linters.

File: tools/test-lib/react-app/jest.setup.js
Comment (limdauto) [inline tools/test-lib/react-app/jest.setup.js:1]: @rashidakanchwala this fixes the `window.URL.createObjectURL` is not a function issue. The other day we were adding it to the wrong place. It was failing inside this test for importing KedroViz in react-app, not at the top level tests.

File: src/components/metadata/metadata.test.js
Comment (bru5) [inline src/components/metadata/metadata.test.js]: I thought these changes were already merged in the parameter hiding branch? Do you need to sync this with main maybe?

File: src/store/index.js
Comment (bru5) [inline src/store/index.js:59]: Unless Prettier is doing this automatically I think ideally try to retain the original whitespace as much as possible, as often newlines are used to help separate logical groups or emphasise certain lines, the smaller change set can also make reviews a bit easier too.

File: src/utils/graph/graph.js
Comment (bru5) [inline src/utils/graph/graph.js:2]: Similar thing here, much appreciated if you can avoid committing whitespace changes in unrelated files (unless Prettier is doing this, we should look into this if so!)

File: src/utils/plot-templates/dark.js
Comment (bru5) [inline src/utils/plot-templates/dark.js:1]: I think best to use a block style comment here and the other file (as just a general style convention for file level docs)

**Review by bru5 — CHANGES_REQUESTED** [review]

File: src/components/metadata/styles/metadata.scss
Comment (studioswong) [inline src/components/metadata/styles/metadata.scss:8]: I have been getting an 'undefined variable' error for the `color-button` variable on running the app from this branch, and indeed I do notice that `color-button` is not defined in `_variables.scss` or anywhere in the app- I wonder if there are any further new changes to the variables scss file that hasn't been pushed?

File: src/components/node-list/node-list-items.js
Comment (studioswong) [inline src/components/node-list/node-list-items.js:81]: I wonder why is there an empty row added here? It doesn't seem to relate to this PR and shouldn't be committed.

File: src/components/node-list/node-list-items.test.js
Comment (studioswong) [inline src/components/node-list/node-list-items.test.js:123]: Similar to the above comment, I understand that these changes could be due to some prettier settings on update, but given that there are no changes introduced to the code and it is not related to this PR, please avoid commiting these type of unrelated changes in the PR to contaminate the clarity in the PR.

File: src/components/metadata/styles/metadata.scss
Comment (rashidakanchwala) [reply src/components/metadata/styles/metadata.scss:8]: thanks for highlighting. this is fixed

File: src/components/plotly-chart/index.js
Comment (studioswong) [reply src/components/plotly-chart/index.js:16]: I think Liam meant here is to change the `hideToolBar` variable to a more general variable name such as `plotConfig`, which can be done here as it's only used as a variable to be passed into the plotly component.

Comment (rashidakanchwala) [reply src/components/plotly-chart/index.js:16]: yes, i have done exactly in the final PR. where it has to displayModeBar for modal but hide it for preview. I have done a slightly different implementation in that but used plotConfig instead of specific name

File: src/reducers/nodes.js
Comment (studioswong) [reply src/reducers/nodes.js:63]: I am still confused as to why `datasetType` is removed here as I recall this field is particularly returned for `data` type nodes ( which `data` type node response object with the `type` field, it is different from the general node.type field that we use within the app.) 

I just checked and can confirm that the response object still returns the field `type` for data objects ( see below), hence there is a need to keep the `datasetType` field. 
image

File: src/reducers/reducers.test.js
Comment (studioswong) [inline src/reducers/reducers.test.js:1]: any particular reason as to why we are changing the naming here? I think maintaining snake case to follow the name of the file would allow this variable to be more untuitive to imitate a data response.

File: src/selectors/nodes.js
Comment (studioswong) [inline src/selectors/nodes.js:130]: In referencing that the plot data is saved under `node.plot`, we should use that field to reference whether the icon is a plot icon rather than using `datasetType`?

File: src/selectors/nodes.js
Comment (studioswong) [inline src/selectors/nodes.js:266]: same comment as above

File: src/store/__mocks__/load-data.js
Comment (studioswong) [inline src/store/__mocks__/load-data.js:4]: see above comment about the casing - the naming should maintain as a snake case to follow the naming of the file

File: src/store/index.js
Comment (studioswong) [reply src/store/index.js:59]: I agree with @bru5 , and on top of that this shouldn't be commited as part of this PR. @rashidakanchwala we can go through your set up to ensure these unrelated changes in files that are unrelated to this PR doesn't get committed

File: src/store/__mocks__/load-data.js
Comment (rashidakanchwala) [reply src/store/__mocks__/load-data.js:4]: @bru5  asked me to change to the camel case. I can't reference his comment anymore :S ... can't find it.

File: src/selectors/nodes.js
Comment (rashidakanchwala) [reply src/selectors/nodes.js:130]: @studioswong  -- as per Lim's new back-end. this works.

File: src/reducers/reducers.test.js
Comment (rashidakanchwala) [reply src/reducers/reducers.test.js:1]: @studioswong  - @bru5 asked me to make it camel case. @bru5 can you comment why and if it's ok to revert?

File: src/selectors/nodes.js
Comment (rashidakanchwala) [reply src/selectors/nodes.js:266]: @studioswong -- as per Lim's new back-end. this works for now as MVP

File: src/store/__mocks__/load-data.js
Comment (studioswong) [reply src/store/__mocks__/load-data.js:4]: aargh I see - yea tbh this change request is out of scope for this PR, and now I see where this is coming from - this is worth a discussion, but would be out of scope for this PR and we can save it for another time.

**Review by studioswong — APPROVED** [review]
approving this to move this forward to the next PR - I haven't gotten the chance to test this on a real kedro project against the server, but as @limdauto mentioned, given that this feature will not be exposed unless the user has a specific setup, it is fine to merge in :) Will be  doing the main testing with an actual kedro project in the next PR.

It would be great to remove the out of scope changes ( i.e the change from snake case to camel case) and change this in a separate PR if needed to prevent confusion down the line. 

Thank you for the work!

File: src/utils/graph/graph.js
Comment (bru5) [reply src/utils/graph/graph.js:2]: Ah no problem, looks like there are some Prettier version conflicts we'll need to investigate later.

File: src/reducers/reducers.test.js
Comment (bru5) [reply src/reducers/reducers.test.js]: Yeah I raised this since the snake case naming style here seems to go against the standards followed in the rest of the project and it wasn't clear as to why, so I thought it had been by mistake.

We have recently written the project standards down in the [style guide](https://github.com/quantumblacklabs/kedro-viz/blob/main/STYLE_GUIDE.md) and if there's a good reason to we can update it, but we should really try avoid changing things too often or adding special cases as it defeats the purpose.

Anyway since this is already snake case in main, we should probably address any reasoning behind this approach later then.

File: src/components/metadata/index.js
Comment (bru5) [inline src/components/metadata/index.js:88]: I think this `pipeline-metadata__list` now needs to change from `<dl>` to `<div>` and then change the `pipeline-metadata__properties` to be the`<dl>` to maintain the existing semantics

**Review by bru5 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
